### ISSUE TITLE
OOB OAuth just got turned off

### ISSUE THREAD

**Issue by nhorton** [GoogleCloudPlatform/docker-credential-gcr#121]
login still uses OOB and see https://developers.google.com/identity/protocols/oauth2/resources/oob-migration

**Comment by jonjohnsonjr** [comment]
It seems like this only affects `docker-credential-gcr gcr-login` invocations that set `-no-browser`?

And there's no other alternative for desktop applications (if I'm reading [this](https://developers.google.com/identity/protocols/oauth2/native-app#manual-copypaste) right) besides using the browser flow.

Would you suggest we just drop this flag? Or is there something else we could do here?

**Comment by lindsayismith** [comment]
We're going through the process of getting this oauth client as onboarded as 'official' on the google side. Does gcr-login not work at all now?

**Comment by lindsayismith** [comment]
(Internal only) screenshot of the challenge screen:

https://screenshot.googleplex.com/38kwQYgeKABsACx

**Comment by rschwarz-sotec** [comment]
Any update on this? Will it be fixed? Or does anybody have an alternative approach how to get this functionality?

**Comment by lindsayismith** [comment]
We are working on getting the OAuth flow approved.  Can you please provide exactly what commands you are running that have changed?  We want to make sure we understand all the client interactions that we need to check.

**Comment by rschwarz-sotec** [comment]
We are using `docker-credential-gcr configure-docker` and `docker-credential-gcr gcr-login` to authenticate developers with their personal accounts on our iot edge devices. This is part of our development and support workflow.

**Comment by lindsayismith** [comment]
Just to make sure we're aligned - currently `docker-credential-gcr configure-docker` works as it used to, and `docker-credential-gcr gcr-login` also works in my testing.

But `docker-credential-gcr gcr-login -no-browser` is broken.

If this isn't the same as you're seeing, please let us know.

**Comment by matthiasfeurer** [comment]
Yes, `docker-credential-gcr configure-docker` works as it used to.
`docker-credential-gcr gcr-login` does not work any more with and without `-no-browser` option.

**Comment by lindsayismith** [comment]
`docker-credential-gcr gcr-login` doesn't any more work if it is unable to launch the browser.  In this case it works the same as -no-browser.

We are working on fixing the oauth flow.

There are some potential workarounds, they involve using gcloud to do authentication.  I know that's not an option for every situation.

If gcloud is installed, docker-credential-gcr will fall back to using gcloud for authentication if you don't do gcr-login. 

Alternatively you can use gcloud directly for docker authentication - see 'gcloud auth configure-docker'

**Comment by gyre** [comment]
Would some variant of this work perhaps?

[https://stackoverflow.com/questions/71364188/how-to-authorize-a-curl-script-to-google-oauth-after-oauth-out-of-band-oob-flo/72981466#72981466](url)

**Comment by matthiasfeurer** [comment]
> Would some variant of this work perhaps?
> 
> https://stackoverflow.com/questions/71364188/how-to-authorize-a-curl-script-to-google-oauth-after-oauth-out-of-band-oob-flo/72981466#72981466

I did a quick test of it and ran into the same problem...

**Comment by lindsayismith** [comment]
Yes we are working on this. To have the oauth flow trusted we have to go through a security review, and we're working through that right now.

**Comment by gyre** [comment]
That's good, thanks!

Do you have a possible timescale for this?  Days?  Weeks? Months?

Thanks!

**Comment by FranciscoKurpiel** [comment]
Not good at all. This attempt to improve security will end up with lots of people using solutions that have worst security properties, like the service json key, or even worst. I can't think on an alternative that has acceptable usability and good security properties. Even if I want to write custom code to generate and refresh json keys, there is till the problem of the root of trust for the rotation. This whole thing is counter-productive and punishes people who adopted gcr/gar.
